lib/AST/ProtocolConformance.cpp

```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
 
   return result;
 }
+
+/// Check of all types used by the conformance are canonical.
+bool ProtocolConformance::isCanonical() const {
+  // Normal conformances are considered canonical by
+  // construction.
+  if (getKind() != ProtocolConformanceKind::Normal) {
+    if (!getType()->isCanonical())
+      return false;
+    if (!getInterfaceType()->isCanonical())
+      return false;
+  }
+
+  switch (getKind()) {
+  case ProtocolConformanceKind::Normal: {
+    auto normalConformance = cast<NormalProtocolConformance>(this);
+    auto sigConformances = normalConformance->getSignatureConformances();
+    for (auto conf : sigConformances)
+      if (!conf.isCanonical())
```

Comment:
I think we want to define `NormalProtocolConformance` as always being canonical, because they are unique; it doesn't matter that we might have type sugar in the signature conformances.